Mathlib/Analysis/Calculus/ImplicitContDiff.lean

```diff
@@ -0,0 +1,141 @@
+that the implicit function preserves the smoothness class of the implicit equation.
+
+Let `E` and `F` be real or complex Banach spaces. Let `f : E × F → F` be a function that is $C^n$ at
+a point `(a, b) : E × F`, where `n ≥ 1`. Let `f'` be the derivative of `f` at `(a, b)`. If the range
+of `f'` is all of `F`, and the kernel of `f'` is the subspace `E × {0}` in `E × F`, then there
+exists a function `φ : E → F` such that `φ a = b`, and `f x (φ x) = f a b` holds for all `x` in a
+neighbourhood of `a`. Furthoremore, `φ` is $C^n$ at `a`.
+
+## TODO
+* Local uniqueness of the implicit function
+* Derivative of the implicit function
+
+## Tags
+
+implicit function, inverse function
+-/
+
+namespace ImplicitFunctionData
+
+variable {𝕜 : Type*} [RCLike 𝕜] {E : Type*} [NormedAddCommGroup E]
+  [NormedSpace 𝕜 E] [CompleteSpace E] {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F]
+  [CompleteSpace F] {G : Type*} [NormedAddCommGroup G] [NormedSpace 𝕜 G] [CompleteSpace G]
+  (φ : ImplicitFunctionData 𝕜 E F G) {n : WithTop ℕ∞}
+
+theorem contDiff_implicitFunction (hl : ContDiffAt 𝕜 n φ.leftFun φ.pt)
+    (hr : ContDiffAt 𝕜 n φ.rightFun φ.pt) (hn : 1 ≤ n) :
+    ContDiffAt 𝕜 n φ.implicitFunction.uncurry (φ.prodFun φ.pt) := by
+  rw [implicitFunction, Function.uncurry_curry, toOpenPartialHomeomorph,
+    ← HasStrictFDerivAt.localInverse_def]
+  exact (hl.prodMk hr).to_localInverse (φ.hasStrictFDerivAt |>.hasFDerivAt) hn
+
+end ImplicitFunctionData
+
+open Filter
+
+open LinearMap (ker range)
+
+open scoped Topology
+
+/-- A predicate stating the sufficient conditions on an implicit equation `f : E × F → F` that will
+lead to a $C^n$ implicit function `φ : E → F`. -/
+structure IsContDiffImplicitAt {𝕜 : Type*} [RCLike 𝕜]
+    {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+    {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F] [CompleteSpace F]
+    (n : WithTop ℕ∞) (f : E × F → F) (f' : E × F →L[𝕜] F) (a : E × F) : Prop where
+  hasFDerivAt : HasFDerivAt f f' a
+  contDiffAt : ContDiffAt 𝕜 n f a
+  range_eq_top : range f' = ⊤
+  ker_eq_left : ker f' = range (LinearMap.inl 𝕜 E F)
+  one_le : 1 ≤ n
+
+namespace IsContDiffImplicitAt
+
+variable
+  {𝕜 : Type*} [RCLike 𝕜]
+  {E : Type*} [NormedAddCommGroup E] [NormedSpace 𝕜 E] [CompleteSpace E]
+  {F : Type*} [NormedAddCommGroup F] [NormedSpace 𝕜 F] [CompleteSpace F]
+  {n : WithTop ℕ∞} {f : E × F → F} {f' : E × F →L[𝕜] F} {a : E × F}
+
+/-- We record the parameters of our specific case in order to apply the general implicit function
+theorem. -/
+def implicitFunctionData (h : IsContDiffImplicitAt n f f' a) :
+    ImplicitFunctionData 𝕜 (E × F) E F where
+  leftFun := Prod.fst
+  leftDeriv := ContinuousLinearMap.fst 𝕜 E F
+  rightFun := f
+  rightDeriv := f'
+  pt := a
+  hasStrictFDerivAt_leftFun := by fun_prop
+  hasStrictFDerivAt_rightFun := h.contDiffAt.hasStrictFDerivAt' h.hasFDerivAt h.one_le
```

Comment:
Should `ContDiffAt.hasStrictFDerivAt'` become a fun_prop lemma? What do you think? (This can happen in a follow-up PR.)

Reply:
I don't have any intuition for when a lemma should be in `fun_prop`. It works like magic to me, when it does. Perhaps you have a better idea.

Reply:
In any case, that probably wouldn't suffice to make this line into a proof by fun_prop --- so let's leave that discussion to another time.